### DISCUSSION TITLE
How to call openai api in client component?

### DISCUSSION THREAD

**Discussion by alamenai** [vercel/next.js#51458] [App Router] [answered]
I just started using openai api ( GPT-3 )  and I'm facing issues related fetching.

### Reproduce

1. Generate Key from the openai platform.
2. Put the Key in `.env` file.
3. Fetch the data from a `Button` component.

### Code

```typescript

'use client';

import Image from 'next/image';
import { useRequestStore } from '../../Request/_state';
import { usePromptStore } from '../_state/store';
import { Configuration, OpenAIApi } from 'openai';
const configuration = new Configuration({
	apiKey: 'My-key'
});

// Delete it
delete configuration.baseOptions.headers['User-Agent'];

const openai = new OpenAIApi(configuration);

export const Button = () => {
	//
	const setRequests = useRequestStore.getState().setRequests;
	const prompt = usePromptStore.getState().prompt;

	const sendRequest = async (event: React.MouseEvent<HTMLImageElement>) => {
		const request = {
			title: prompt,
			output: 'Done',
		};
		setRequests(request);

		const chat_completion = await openai.createChatCompletion({
			model: 'gpt-3.5-turbo',
			messages: [{ role: 'user', content: 'Hello world' }],
		});

		console.log(chat_completion); // expect results
	};

	return (
		<Image
			src={'/send.svg'}
			width={40}
			height={40}
			alt='send request icon'
			className='bg-black p-3 rounded-full cursor-pointer hover:shadow-lg z-30'
			onClick={sendRequest}
			role='button'
		/>
	);
};

export default Button;

```

### Expected

Log the results

### Current

```
createError.js?d682:16 Uncaught (in promise) Error: Request failed with status code 429
    at createError (webpack-internal:///(:3000/app-client)/./node_modules/axios/lib/core/createError.js:16:15)
    at settle (webpack-internal:///(:3000/app-client)/./node_modules/axios/lib/core/settle.js:17:12)
    at XMLHttpRequest.onloadend (webpack-internal:///(:3000/app-client)/./node_modules/axios/lib/adapters/xhr.js:66:7)
```

### Screenshot

![Screenshot from 2023-06-18 00-05-24](https://github.com/vercel/next.js/assets/20143684/c7b3ba39-41b9-4b60-9f9e-995fc9752de0)

**Comment by icyJoseph** [comment, marked as answer]
Hi,

First, careful with sharing your API keys in public forums like this. Is that key supposed to be public?

The openai npm package states:

> ⚠️ Important note: this library is meant for server-side usage only, as using it in client-side browser code will expose your secret API key. [See here](https://platform.openai.com/docs/api-reference/authentication) for more details.

Anyway, using the key you shared, I get that the key is invalid, so it was probably deleted. You gotta generate a new one, and NOT share it!

Second, 401 (from the screenshots) means authorised, so perhaps your key can't use that API at all?

Third, 429, means too many requests, so I guess, somehow the error chained into many requests, and the API throttled you down?

Last, but not least, https://stackoverflow.com/a/75950173 If you go to `/account/usage` on the Open AI website, you'll see, how much credit you've got left.
